File: sast-python2/case/completeness/single_app_tracing/function_call/higher_order_function/higher_order_function_008_T.py
# -*- coding: utf-8 -*-
# evaluation information start
# real case = true
# evaluation item = 完整度->单应用跟踪完整度->函数和方法调用->高阶函数
# scene introduction = 匿名函数为参数
# level = 2
# bind_url = completeness/single_app_tracing/function_call/higher_order_function/higher_order_function_008_T
# evaluation information end
import os


def higher_order_function_008_T(taint_src):
    def higher_order_function(callback):
        return callback(taint_src, u'_')

    result = higher_order_function(lambda a, b: a + b)
    taint_sink(result)


def taint_sink(o):
    os.system(o)


if __name__ == u'__main__':
    taint_src = u"taint_src_value"
    higher_order_function_008_T(taint_src)

File: sast-js/case/completeness/single_app_tracing/function_call/higher_order_function/higher_order_function_008_T.js
// evaluation information start
// real case = true
// evaluation item = 完整度->单应用跟踪完整度->函数和方法调用->高阶函数
// scene introduction = 箭头函数是参数
// level = 2
// bind_url = completeness/single_app_tracing/function_call/higher_order_function/higher_order_function_008_T
// evaluation information end
const { execSync } = require('child_process');


function higher_order_function_008_T(__taint_src) {
  function higherOrderFunction(callback) {
    return callback(__taint_src, "_");
  }

  const result = higherOrderFunction((a, b) => a + b);
  __taint_sink(result);
}

function __taint_sink(o) {
  execSync(o);
}

const taint_src = "taint_src_value";

higher_order_function_008_T(taint_src);
